kxclib.ml

```diff
@@ -1907,3 +1907,192 @@ module Jv = struct
+        iofs, ilen, itill
+      else invalid_arg "Base64.decode: wrong padding"
+    in
+    let output =
+      let estimated_bytes = (ilen / 4) * 3 + 2 in  (* worst case: 3n+2 bytes (= 4n+3 chars) *)
+      create estimated_bytes
+    in
+    let read stack o =
+      let set o value = set_uint8 output o (value land 0xff); o+1 in
+      match List.rev stack with
+      | [] -> o
+      | [_] -> invalid_arg "Base64.decode: unterminated input"
+      | s1 :: s2 :: ss ->
+        let o = set o ((s1 lsl 2) lor (s2 lsr 4)) in (* s1[0]..s1[5] s2[0] s2[1] *)
+        match ss with
+        | [] -> (* [3n+1 bytes] 4bits = s2[2]..s2[5] should've been padded *)
+          if not ((s2 land 0xf) = 0) then invalid_arg "Base64.decode: unterminated input"
+          else o
+        | s3 :: ss ->
+          let o = set o ((s2 lsl 4) lor (s3 lsr 2)) in (* s2[2]..s1[5] s3[0]..[3] *)
+          match ss with
+          | [] -> (* [3n+2 bytes] 2bits = s3[4] s3[5] should've been padded *)
+            if not ((s3 land 0x3) = 0) then invalid_arg "Base64.decode: unterminated input"
+            else o
+          | s4 :: [] -> (* [3n bytes] *)
+            set o ((s3 lsl 6) lor s4) (* s3[4] s3[5] s4[0]..[5] *)
+          | _ -> failwith "impossible"
+    in
+    let check_padding s =
+      match String.length s with
+      | slen when slen > 0 && slen <= 2 ->
+         iotafl slen (fun acc i -> acc && String.get s i = pad_char) true
+      | _ -> false in
+    let rec go stack i o =
+      if i = itill then read stack o
+      else
+        let c = get_uint8 input i in
+        match char_to_sixbit c with
+        | None ->
+          begin match char_of_int c with
+          | _ when ignore_unknown -> go stack (i+1) o
+          | '\r' | '\n' when ignore_newline -> go stack (i+1) o
+          | c when c = pad_char && not no_padding ->
+            let rest = sub_string input i (itill - i) in
+            if check_padding rest then read stack o
+            else invalid_arg' "Base64.decode: invalid char '%c' at index %d" pad_char i
+          | c -> invalid_arg' "Base64.decode: invalid char '%c' at index %d" c i
+          end
+        | Some s ->
+          let stack = s :: stack in
+          if List.length stack = 4 then
+            let o = read stack o in
+            go [] (i+1) o
+          else
+            go stack (i+1) o
+    in
+    let actual_bytes = go [] iofs 0 in
+    Bytes.sub output 0 actual_bytes
+
+  type base64_codec = (?sub:int * int -> bytes -> string) * (?sub:int * int -> string -> bytes)
+
+  let base64_codec ?no_padding ?c62 ?c63 ?pad_char ?prefix
+        ?ignore_newline ?ignore_unknown
+        () : base64_codec =
+    let enc = encode ?no_padding ?c62 ?c63 ?pad_char ?prefix in
+    let dec = decode ?no_padding ?c62 ?c63 ?pad_char ?prefix ?ignore_newline ?ignore_unknown in
+    enc, dec
+
+  let encode_rfc4648' = encode ~no_padding:false ~c62:'+' ~c63:'/' ~pad_char:'='
+  let decode_rfc4648' ?(ignore_newline=true) =
```

Comment:
rfc4846 の ignore_newline の default を true にしました

Reply:
これは仕様違反

> Implementations MUST reject the encoded data if it contains
   characters outside the base alphabet when interpreting base-encoded
   data, unless the specification referring to this document explicitly
   states otherwise.
> Such specifications may instead state, as MIME
   does, that characters outside the base encoding alphabet should
   simply be ignored when interpreting data ("be liberal in what you
   accept").  Note that this means that any adjacent carriage return/
   line feed (CRLF) characters constitute "non-alphabet characters" and
   are ignored.